Rubyプラクティス > lsコマンドを作る2 として -aオプション機能を追加した

mochiz opened the pull request:
## 背景
- #4
- で作成した ls.rb に ls -a 相当のオプション機能を追加したい

## やったこと
- [x] -a オプションを指定してファイル名一覧を表示できるようにした
- [x] items という変数名を file_names に差し替えた

## 確認方法

ディレクトリに移動して ./ls.rb - aでファイル一覧を表示できること

```sh
$ cd 04.ls
$ ./ls.rb -a
```

- 実行結果
  - ![CleanShot 2024-10-17 at 12 09 52@2x](https://github.com/user-attachments/assets/1e16cfc0-4cc5-463b-8d88-271d146b0f1c)
- lsコマンドとの比較
  - ![CleanShot 2024-10-17 at 11 51 05@2x](https://github.com/user-attachments/assets/f4f606c8-1b44-4f70-94b7-b9055fb56888)
- rubocop実行結果
  - ![CleanShot 2024-10-17 at 12 06 46@2x](https://github.com/user-attachments/assets/b6e7a8f3-52a2-43e6-8ecc-a2c2b2fb0270)

mochiz committed:
item という変数名を file_name に差し替えた

mochiz committed:
ls.rb に -a オプション機能を追加した

mochiz committed:
sorted_file_names は冗長なので file_names に変更した

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,22 +1,33 @@
+  if options['a']
+    file_names.sort
+  else
+    file_names.sort.reject { |file_name| file_name.start_with?('.') }
```
`file_names.sort`がDRYじゃないですね

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,22 +1,33 @@
-sorted_items = items.sort.reject { |item| item.start_with?('.') }
+options = ARGV.getopts('a')
+
+def get_file_names(options)
```
9行目のoptionsと名前が同じで紛らわしいですね。
そもそもトップレベルのロジックとメソッドのロジックを混在させると混乱の元なので、mainメソッドを導入してみてください。
https://qiita.com/jnchito/items/4b4cae54170cc2f4377e

JunichiIto replied:
あと、Rubyでget_ や set_ で始まるメソッドはあまり使わないので、get以外の動詞を持ってくる方がベターです。

JunichiIto reviewed (commented):
コメントしました〜。

mochiz committed:
ls.rbについて指摘箇所を修正した

- optionsが紛らわしいので ARGVで取得したオプションはargv_optionsと別名にした
- get_ prefixはrubyらしくないので削除した
- sort 処理がDRYじゃないのでsorted_file_names 変数に保持するよう修正した

mochiz commented:
レビューありがとうございます！指摘いただいた箇所を調整してみました。ちょっと自信ないですがどうでしょうか
お手すきの際に再レビューお願いします 🙇

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -6,28 +6,34 @@
-  end
-  puts
-end
+main(argv_options)
```
```suggestion
main
```

argv_optionsだけmainの外に出した理由って何かありますか？
それも含めて全部mainの中でやっちゃった方がいいのでは？

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -6,28 +6,34 @@
 
-def get_file_names(options)
+def main(options = {})
+  file_names = file_names(options)
```
メソッド名と変数名を同じにするのはなるべく避けましょう
https://qiita.com/jnchito/items/fb438aa0786bb3cbf919

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,22 +1,39 @@
-  row_items.each do |item|
-    print item.ljust(padding_size)
+def file_names(options = {})
+  file_names = Dir.entries(__dir__)
```
ここも自分自身のメソッド名と同じ変数名になってますね

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,22 +1,39 @@
-    print item.ljust(padding_size)
+def file_names(options = {})
+  file_names = Dir.entries(__dir__)
+  sorted_file_names = file_names.sort
```
```suggestion
  file_names = Dir.entries(__dir__).sort
```

変数代入の時点でsortしてしまっていいのでは？

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,22 +1,39 @@
+def file_names(options = {})
+  file_names = Dir.entries(__dir__)
+  sorted_file_names = file_names.sort
+  if options['a']
```
三項演算子で書いた方がスッキリ書けるかも？

JunichiIto reviewed (commented):
コメントしました！

mochiz committed:
指摘いただいた箇所を修正した

mochiz commented:
レビューありがとうございます！指摘いただいた箇所を修正しました。スッキリとわかりやすくなったと思います。
何度もすみません。引き続きよろしくお願いします。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -26,14 +25,9 @@ def main(options = {})
-  else
-    sorted_file_names.reject { |file_name| file_name.start_with?('.') }
-  end
+def generate_file_names(options = {})
```
generateだと無から何かを作り出すような印象がありますね。
（パスワードジェネレータ的な）

どちらかというとすでにあるデータを取捨選択して持ってくるイメージなので、searchやfind みたいな名前の方がいいかもしれません。

JunichiIto reviewed (approved):
1点コメントしましたが、ls3提出時に修正してもらったらOKです。
この課題はこれでOKとしますね〜。
